Sha3 256

Matt-Ryley opened the pull request:
1. Added the SHA3-256 OPCode and functionality
2. Added the SHA3-256 to the unit tests
3. Amended the documentation so that SHA3-256 is included

-- The unit tests have been run on my local machine and all pass.

Matt-Ryley committed:
SHA3-256 added

Added SHA3-256 hashing function as an OP code

Matt-Ryley committed:
Unit tests now pass

Made required changes to pass the unit tests.
-- Changed testAccepts version from 1 to 6
-- Added sha3_256 to the Opgroups

tgrbrooks commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1154,6 +1154,15 @@ bitlen interprets arrays as big-endian integers, unlike setbit/getbit
 
+## sha3_256
+
+- Opcode: 0x96
```
I don't know how this was handled previously, but all the crypto related opcodes are next to each other.
They haven't all existed from the start so they must have either left space for them, or shifted all of the other ones down.
Would be worth looking up where some of the new ones were added (e.g. `ecdsa_pk_recover`) and seeing what they did

Matt-Ryley replied:
For the new ecdsa there was a gap in the op_codes which they filled. I see for the PR https://github.com/algorand/go-algorand/pull/3218, which is a new OP code for for ed25519 verification they just select an unused opcode, I think they did the same for the new math opcodes, see the new math codes above the sha3_256

tgrbrooks replied:
Ah ok, this should be good then, thanks for looking!

jannotti replied:
We do try to leave them where they are.  FYI, 0x96 is going to be taken in the next release by `bsqrt`.  I think we're considering the PR for SHA3-256 now, but it would probably be in the following TEAL version.  If compatibility with that matters, maybe move this to 0x9f for now, which saves a little room for us in case another opcode or two goes in.

But then again, perhaps you're just prototyping with this, and you wouldn't mind a change whenever we figure out a spot for this.

Matt-Ryley replied:
Hi, we're happy to wait and then to change the opcode when you have figured out a spot for it.

tgrbrooks commented on data/transactions/logic/eval.go:
```diff
@@ -901,6 +901,16 @@ func opSHA256(cx *EvalContext) {
 }
 
+// The NIST SHA3-256 is implemented for compatability with ICON
+func opSHA3256(cx *EvalContext) {
```
Can this be `opSHA3_256`?

Matt-Ryley replied:
Yes it can and should! Should be resolved in most recent commit

tgrbrooks reviewed (commented):
Looks great, thank you!
Just a couple of things to double check/change but almost good to go

Matt-Ryley committed:
SHA3256->SHA3_256

From code review suggestion renamed opSHA3256 to opSHA3_256 in the eval.go and opcodes.go file

tgrbrooks reviewed (approved):
Great, thank you!
